front: display output table for invalid trains

elise-chin opened the pull request:
Closes https://github.com/osrd-project/osrd-confidential/issues/1146 (1st part)
The new appearance of invalid trains is written in this PR https://github.com/OpenRailAssociation/osrd/pull/12611. 

Reviewers are advised to review this PR commit-by-commit and disable whitespace changes in the diff view.

## Tests

Examples:
[invalid_trains_output_table_timetable.json](https://github.com/user-attachments/files/21544416/invalid_trains_output_table_timetable.json)

### Valid trains
Check that the output table is still well displayed with every expected rows for valid trains schedules and paced trains.
Especially for paced trains, check that the cadencing is updated for each occurrence.

### Invalid trains
Check the output table of:
- Invalid train schedules
  - Without explicit track
  - With explicit track selected for an operational point
  - With waypoints selected directly on the map, that are not operational point
  - With a created NGE node that do not exist on the infrastructure 
- Invalid occurrences: same as above but for every occurrence

elise-chin committed:
front: remove pathStepRows' name in useOutputTableData

It is already defined when we select a waypoint in the map in upsertMapWaypointsInOperationalPoints > baseFormattedStep

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

elise-chin committed:
front: return the train and rollingStock when the pathfinding or simulation has failed in useSimulationResults

We will need the train to display the output table for an invalid train.

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

elise-chin committed:
front: render TimesStopsOutput for valid and invalid trains

Moves rendering of TimesStopsOutput outside of the isValid check, and
keeps SpeedSpaceChart, SimulationMap, and SimulationResultExport components
restricted to valid simulation results only.

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

Caracol3 commented on front/src/applications/operationalStudies/hooks/useInvalidTrainOps.ts:
```diff
@@ -0,0 +1,64 @@
+): PathPropertiesFormatted['operationalPoints'] => {
+  const operationalPointReferences: OperationalPointReference[] = useMemo(() => {
+    if (!simulationResults || simulationResults.isValid) return [];
+    return simulationResults.train.path
+      .filter((pathItem) => isOperationalPointReference(pathItem))
+      .map(
+        ({ id: _id, deleted: _deleted, ...cleanOperationalPointReference }) =>
+          cleanOperationalPointReference
+      );
+  }, [simulationResults]);
```
We could avoid to iterate twice with `filter`an `map`like this
```suggestion
   return simulationResults.train.path.reduce((acc, pathItem) => {
  if (isOperationalPointReference(pathItem)) {
    const { id: _id, deleted: _deleted, ...cleanOperationalPointReference } = pathItem;
    acc.push(cleanOperationalPointReference);
  }
  return acc;
}, [] as OperationalPointReference[]);
```

SharglutDev replied:
@emersion you didn't see this comment ❌

SharglutDev replied:
> [!TIP]
>   To avoid the cast in the `reduce`, you can also do : 
```suggestion
   return simulationResults.train.path.reduce<OperationalPointReference[]>((acc, pathItem) => {
  if (isOperationalPointReference(pathItem)) {
    const { id: _id, deleted: _deleted, ...cleanOperationalPointReference } = pathItem;
    acc.push(cleanOperationalPointReference);
  }
  return acc;
}, []);
```

elise-chin replied:
Done!

Caracol3 reviewed (commented):
Nice work !
Just left one comment.

Caracol3 commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
```diff
@@ -260,7 +267,7 @@ const SimulationResults = ({
       </ResizableSection>
 
-      {simulationResults && (
+      {simulationResults && simulationResults.isValid && (
```
It should be fine like this :

```suggestion
      {simulationResults?.isValid && (
```

elise-chin replied:
Done!

clarani commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
```diff
@@ -291,32 +298,42 @@ const SimulationResults = ({
+            selectedTrain={simulationResults?.train}
+            isValid={isValid}
+            simulatedTrain={isValid ? simulationResults.simulation?.final_output : undefined}
+            simulatedPath={isValid ? simulationResults.path : undefined}
```
Nit ;)
```suggestion
            {...(simulationResults?.isValid
              ? {
                  isValid: true,
                  simulatedTrain: simulationResults.simulation.final_output,
                  simulatedPath: simulationResults.path,
                }
              : { isValid: false })}
```

elise-chin replied:
Aaah, I didn't know how to do that, thanks!

clarani commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
```diff
@@ -260,7 +267,7 @@ const SimulationResults = ({
       </ResizableSection>
 
-      {simulationResults && (
+      {simulationResults && simulationResults.isValid && (
```
`simulationResults?.isValid` is enough ;) 
```suggestion
      {simulationResults?.isValid && (
```

elise-chin replied:
Done!

clarani commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
```diff
@@ -70,6 +71,7 @@ const SimulationResults = ({
 
   const simulationResults = useSimulationResults(infraId);
   const selectedTrainId = simulationResults?.train.id;
+  const isValid = simulationResults?.isValid;
```
You can remove this const and always use `simulationResults?.isValid` directly

elise-chin replied:
Done!

Caracol3 commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
```diff
@@ -260,7 +267,7 @@ const SimulationResults = ({
 
-      {simulationResults && (
+      {simulationResults && simulationResults.isValid && (
         <>
```
We can remove this too

```suggestion
        
```

elise-chin replied:
I removed the one from line 325 😉

Caracol3 commented on front/src/applications/operationalStudies/hooks/useInvalidTrainOps.ts:
```diff
@@ -0,0 +1,64 @@
+          operational_point_references: operationalPointReferences,
+        },
+      },
+      { skip: !infraId || operationalPointReferences.length === 0 }
```
We can remove `infraId` as it's always defined
```suggestion
      { skip: operationalPointReferences.length === 0 }
```

elise-chin replied:
Yes, you're right, thanks!

clarani commented on front/src/applications/operationalStudies/hooks/useInvalidTrainOps.ts:
```diff
@@ -0,0 +1,64 @@
+      { skip: !infraId || operationalPointReferences.length === 0 }
+    );
+
+  // Get this type will facilitate the logic in TimesStopsOutput
+  const formattedOperationalPoints: PathPropertiesFormatted['operationalPoints'] = useMemo(() => {
+    if (
+      !operationalPoints?.related_operational_points ||
+      operationalPoints.related_operational_points.length === 0
+    ) {
+      return [];
+    }
+
+    // We have at least one related operational point
+    return operationalPoints.related_operational_points
+      .filter((ops) => ops.length !== 0) // To remove empty arrays related to invalid step
+      .map((ops) => ({
+        ...getStationFromOps(ops)!,
+        // The following properties will not be used for invalid trains output table data
+        part: { position: 0, track: 'unknown' } as OperationalPointPart,
+        position: 0,
+        weight: null,
+      }));
+  }, [operationalPoints]);
```
You can move this useMemo in the useQuery if you use selectFromResult ;) 
```
const formattedOperationalPoints =
    osrdEditoastApi.endpoints.postInfraByInfraIdMatchOperationalPoints.useQuery(
      {
        infraId,
        body: {
          operational_point_references: operationalPointReferences,
        },
      },
      { skip: !infraId || operationalPointReferences.length === 0, selectFromResult: ({ data })  => {... }}
    );
```
You can have a look in ScenarioHeader

elise-chin replied:
I think I need your help for this one, I tried several code, but I got ts errors everytime 😅 

For example with this snippet:
```typescript
  const { formattedOperationalPoints }: PathPropertiesFormatted['operationalPoints'] =
    osrdEditoastApi.endpoints.postInfraByInfraIdMatchOperationalPoints.useQuery(
      {
        infraId,
        body: {
          operational_point_references: operationalPointReferences,
        },
      },
      {
        skip: !infraId || operationalPointReferences.length === 0,
        selectFromResult: ({ data }) => {
          if (!data?.related_operational_points || data.related_operational_points.length === 0)
            return [];

          return data.related_operational_points
            .filter((ops) => ops.length !== 0)
            .map((ops) => {
              const {
                parts: _parts,
                weight: _weight,
                ...cleanOperationalPoint
              } = getStationFromOps(ops) || {};
              return {
                formattedOperationalPoints: {
                  ...cleanOperationalPoint,
                  // The following properties will not be used for invalid trains output table data
                  part: { position: 0, track: 'unknown' },
                  position: 0,
                  weight: null,
                },
              };
            });
        },
      }
    );
```
    
I had the following error
    
```
    Type 'UseQueryHookResult<QueryDefinition<PostInfraByInfraIdMatchOperationalPointsApiArg, BaseQueryFn<FetchArgs, unknown, ApiError>, "authz" | ... 27 more ... | "work_schedules", PostInfraByInfraIdMatchOperationalPointsApiResponse, "editoastApi", unknown>, { ...; }[]>' is not assignable to type '{ extensions?: OperationalPointExtensions | undefined; id: string; part: OperationalPointPart; position: number; weight: number | null; }[]'.
```

elise-chin replied:
Solved with the help of @clarani ❤️

clarani commented on front/src/applications/operationalStudies/hooks/useInvalidTrainOps.ts:
```diff
@@ -0,0 +1,64 @@
+      .filter((ops) => ops.length !== 0) // To remove empty arrays related to invalid step
+      .map((ops) => ({
+        ...getStationFromOps(ops)!,
+        // The following properties will not be used for invalid trains output table data
+        part: { position: 0, track: 'unknown' } as OperationalPointPart,
+        position: 0,
+        weight: null,
```
Do you think we can omit these props ?

elise-chin replied:
If we omit them, we have to make these attributes optional in `PathPropertiesFormatted['operationalPoints']`.

Caracol3 commented on front/src/modules/timesStops/hooks/useOutputTableData.ts:
```diff
@@ -19,48 +19,52 @@ import { ARRIVAL_TIME_ACCEPTABLE_ERROR } from '../consts';
 
-      const computedArrival = new Date(startDatetime.getTime() + pathItemTimes.final[index]);
+      if ('track' in pathStep) mapWaypointCount += 1;
+      const name = isValid
+        ? matchingOperationalPoint?.extensions?.identifier?.name
+        : getOperationalPointName(matchingOperationalPoint, pathStep, mapWaypointCount, t);
```
We can also find the name in `getOperationalPointName`
```suggestion
      const name = getOperationalPointName(matchingOperationalPoint, pathStep, mapWaypointCount, t);
```

elise-chin replied:
Done!

clarani commented on front/src/modules/timesStops/helpers/computeMargins.ts:
```diff
@@ -70,6 +68,15 @@ function computeMargins(
+      calculatedMargin: undefined,
+      diffMargins: undefined,
+    };
+  }
```
You can move this block of code above the for loop which looks for the next pathStep where constraints are defined

elise-chin replied:
Done!

Caracol3 commented on front/src/modules/timesStops/helpers/scheduleData.ts:
```diff
@@ -26,6 +26,13 @@ export const formatSchedule = (arrivalTime: Date, schedule?: ScheduleEntry) => {
+  }
   return {
     stopFor: `${stopFor.total('second')}`,
     calculatedDeparture: dateToHHMMSS(addDurationToDate(arrivalTime, stopFor)),
```
```suggestion
calculatedDeparture:  arrivalTime ? dateToHHMMSS(addDurationToDate(arrivalTime, stopFor)) : undefined,
```

elise-chin replied:
At this step, we know that `arrivalTime` is defined since we checked before at line 29 😄

Caracol3 commented on front/src/modules/timesStops/helpers/utils.ts:
```diff
@@ -341,3 +346,46 @@ export function receptionSignalToSignalBooleans(receptionSignal?: ReceptionSigna
+
+export const getTrackReferenceLabel = (trackReference?: TrackReference | null) => {
+  if (!trackReference) return undefined;
+  if ('track_id' in trackReference) return trackReference.track_id;
+  return trackReference.track_name;
```
```suggestion
  ('track_id' in trackReference) ? trackReference.track_id : trackReference.track_name;
```

elise-chin replied:
Done!

clarani commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
```diff
@@ -144,6 +146,11 @@ const SimulationResults = ({
     return exception?.summary ?? pacedTrain.summary;
   }, [timetableItemsWithDetails, selectedTrainId]);
 
+  const invalidTrainOperationalPoints = useInvalidTrainOps(infraId, simulationResults);
```
Last small comment, could you move this hook in useOutputData since invalidTrainOperationalPoints is only used there ? :) Thanks :pray:

elise-chin replied:
Done!

Caracol3 reviewed (approved):
LGTM

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.ts:
```diff
@@ -94,70 +126,95 @@ const useOutputTableData = (
+            trackName = trackSections[matchingPathStep.track]?.extensions?.sncf?.track_name;
+          } else {
+            trackName = getTrackReferenceLabel(matchingPathStep.track_reference);
+          }
```
nit
```suggestion
         const trackName =
            'track' in matchingPathStep
              ? trackSections[matchingPathStep.track]?.extensions?.sncf?.track_name
              : getTrackReferenceLabel(matchingPathStep.track_reference);
```

elise-chin replied:
Done!

SharglutDev commented on front/src/modules/timesStops/TimesStopsOutput.tsx:
```diff
@@ -14,45 +14,53 @@ import TimesStops from './TimesStops';
+  infraId: number;
   selectedTrain?: Train;
-  path?: PathfindingResultSuccess;
+  isValid?: boolean;
```
`isValid` is always defined (same in `useOutputTableData`)
```suggestion
  isValid: boolean;
```

elise-chin replied:
You're right, done!

SharglutDev commented on front/src/modules/timesStops/hooks/useOutputTableData.ts:
```diff
@@ -20,49 +21,59 @@ import { ARRIVAL_TIME_ACCEPTABLE_ERROR } from '../consts';
       selectedTrain.path.map((pathStep, index) => {
-        const schedule: ScheduleEntry | undefined = scheduleByAt[pathStep.id];
+        const opPositionOnPath = simulatedPath?.path_item_positions[index];
+        const matchingOperationalPoint = operationalPoints.find((op) =>
```
I have a white screen here when entering a scenario

<img width="635" height="54" alt="Capture d’écran 2025-08-07 à 14 58 16" src="https://github.com/user-attachments/assets/06f914a6-0c06-4fa2-b0c3-dea37d6040a1" />

This comes from the way we use `selectFromResult` in `useInvalidTrainOps.ts`. We should always return the `response` alongside the `data`.

Changing the api call with that seems to fix it : 

```
const { data: operationalPoints } =
    osrdEditoastApi.endpoints.postInfraByInfraIdMatchOperationalPoints.useQuery(
      {
        infraId,
        body: {
          operational_point_references: operationalPointReferences,
        },
      },
      {
        skip: operationalPointReferences.length === 0,
        selectFromResult: ({ data, ...response }) => {
          if (!data?.related_operational_points || data.related_operational_points.length === 0)
            return {
              ...response,
              data: [],
            };

          return {
            ...response,
            data: data.related_operational_points
              .filter((ops) => ops.length !== 0)
              .map((ops) => ({
                ...getStationFromOps(ops)!,
                // The following properties will not be used for invalid trains output table data
                part: { position: 0, track: 'unknown' },
                position: 0,
                weight: null,
              })),
          };
        },
      }
    );
```

clarani replied:
Yep, I found this bug too :)

But even with your suggestion, I noticed that the hooks from useOutputData are being triggered in a loop, and it seems to be caused by selectFromResult. I suggest we revert this change for now and avoid using selectFromResult.

elise-chin replied:
Done!

clarani commented on front/src/modules/timesStops/hooks/useOutputTableData.ts:
```diff
@@ -20,49 +21,59 @@ import { ARRIVAL_TIME_ACCEPTABLE_ERROR } from '../consts';
   const scheduleByAt: Record<string, ScheduleEntry> = keyBy(selectedTrain?.schedule, 'at');
   const theoreticalMargins = selectedTrain && getTheoreticalMargins(selectedTrain);
+  const invalidTrainOperationalPoints = useInvalidTrainOps(infraId, selectedTrain, isValid);
+  const operationalPoints = isValid ? simulatedOperationalPoints : invalidTrainOperationalPoints;
```
Could you add a useMemo for operationalPoints ?

Otherwise, at each render, operationalPoints is re-instantiated, so the useMemo and the useEffect are fired again and again

SharglutDev replied:
Still think it would be better to call `useInvalidTrainOps` in `SimulationResults` :)
The output table cannot work without `operationalPoints` so it makes sense to do the `isValid` check over there and always pass an array to the component.

elise-chin replied:
`TimesStopsOutput` does not work when there is no selected train (`dataIsLoading={!selectedTrain}`). I removed the condition on `operationalPoints`.

SharglutDev replied:
Ok let's keep it like that for now :)

clarani reviewed (commented):
Almost good to me !

SharglutDev reviewed (commented):
After fixing the white screen, I now sometimes have a infinite loop : 
<img width="346" height="189" alt="Capture d’écran 2025-08-07 à 15 23 30" src="https://github.com/user-attachments/assets/abb0c0a1-9fbf-49e7-839f-f66031f5414f" />

The timetable I use : 
[timetable_outputtable.json](https://github.com/user-attachments/files/21666194/timetable_outputtable.json)

edit : ok related to @clarani [comment](https://github.com/OpenRailAssociation/osrd/pull/12499#discussion_r2260283484) :)

elise-chin committed:
front: rename props of TimesStopsOutput and useOutputTableData

It is for clearer distinction between simulation-derived and
non simulation props.

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

elise-chin committed:
front: move injection of operational points to path steps rows formatting in useOutputTableData

For better contextual grouping.

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

elise-chin committed:
front: introduce useInvalidTrainOps()

This will be useful to display invalid train's operational points in the output table.

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

elise-chin committed:
front: move getInvalidStepLabel() to operationalStudies/utils.ts

This will be use in timesStops helpers functions.

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

elise-chin committed:
front: introduce matchOpRefAndOp(), getTrackReferenceLabel(), and getOperationalPointName()

These helpers functions will be use for invalid trains' output table.

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

elise-chin committed:
front: make pathItemTimes optional in computeMargins() to accommodate for invalid train's output table

This is to avoid formatting theoreticalMargin again outside this function
in useOutputTableData().

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

elise-chin committed:
front: make arrivalTime optional in formatSchedule() to accomodate for invalid train's output table

This is to avoid formatting stopFor, shortSlipDistance, and onStopSignal again
outside this function in useOutputTableData().

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

elise-chin committed:
front: display invalid's train output table

Calls useInvalidTrainOps(), matches operational point on their reference instead of position,
ignores computed attributes that are specific to valid trains,
and fills the path steps rows with track name.

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

elise-chin committed:
front: compute arrivalScheduleNotRespected and negativeDiffMargins only for valid trains

Signed-off-by: Elise Chin <c.chin.elise@gmail.com>

SharglutDev commented on front/src/modules/timesStops/hooks/useOutputTableData.ts:
```diff
@@ -20,49 +21,62 @@ import { ARRIVAL_TIME_ACCEPTABLE_ERROR } from '../consts';
   const scheduleByAt: Record<string, ScheduleEntry> = keyBy(selectedTrain?.schedule, 'at');
   const theoreticalMargins = selectedTrain && getTheoreticalMargins(selectedTrain);
+  const invalidTrainOperationalPoints = useInvalidTrainOps(infraId, isValid, selectedTrain);
+  const operationalPoints = useMemo(
```
Sometimes (only for invalid trains), the OP names in the table is the uic and not the name. Switching the selection on another train and switching back refreshes the name (probably some missing rerender somewhere)


https://github.com/user-attachments/assets/87b69019-291b-4a4b-9d91-9209a8773e37

clarani replied:
Yes noticed too ;) To fix this, you need to add the invalidTrainOperationalPoints in the dependencies of this useMemo :) 

(Otherwise, if your train is invalid, operationalPoints will be `[]` as simulatedOperationalPoints will be undefined, and won't be recomputed if invalidTrainOperationalPoints changes

elise-chin replied:
Thank you for noticing that and for the explanation, I better understand the logic :D

clarani commented on front/src/applications/operationalStudies/views/SimulationResults.tsx:
```diff
@@ -291,34 +291,45 @@ const SimulationResults = ({
+              : { isValid: false })}
+          />
+        </div>
+      </BoardWrapper>
```
The boardWrapper needs to be after the simulation export buttons to follow the mock-ups ;)

clarani reviewed (approved):
LGTM :white_check_mark: Nice work !

You can resolve my last comments once you have fixed them

SharglutDev reviewed (approved):
Lgtm and tested, great job !
